### PR TITLE
check that LHS are well-formed + eta-expand legacy pattern variables

### PR DESCRIPTION
**Pull request by fblanqui**

### CODE REVIEW

File: src/menhir_parser.mly [hunk 1]
```diff
@@ -59,31 +60,30 @@ let translate_old_rule : old_p_rule -> p_rule = fun r ->
...
                 fatal a.pos "Annotation in legacy pattern."
             | [([Some x],None   ,_)] ->
                 compute_arities (x.elt::env) t
-            | _                      ->
-                fatal h.pos "Invalid legacy pattern lambda."
+            | _                      -> assert false
           end
-      | P_Patt(_,_)     -> fatal h.pos "Pattern in legacy rule."
       | P_Impl(_,_)     -> fatal h.pos "Implication in legacy pattern."
-      | P_LLet(_,_,_,_) -> fatal h.pos "Let expression in legacy rule."
```
Comment (gabrielhdt) [inline src/menhir_parser.mly]: There is a let in legacy Dedukti, so perhaps we should keep the error message for it

Comment (fblanqui) [reply src/menhir_parser.mly]: fixed in c73b6ac

**Comment by fblanqui** [comment]
Why applying a pattern in a RHS would be a problem?

**Comment by gabrielhdt** [comment]
No indeed, I made a mistake.

**Comment by fblanqui** [comment]
Hello @rlepigre . Do you agree with this PR?

File: src/menhir_parser.mly [hunk 2]
```diff
@@ -59,31 +60,30 @@ let translate_old_rule : old_p_rule -> p_rule = fun r ->
             begin
               try
                 let n = Hashtbl.find arity x in
-                if nb_args < n then Hashtbl.replace arity x nb_args
+                if nb_args > n then Hashtbl.replace arity x nb_args
               with Not_found -> Hashtbl.add arity x nb_args
             end
       | P_Wild          -> ()
-      | P_Type          -> ()
```
Comment (rlepigre) [inline src/menhir_parser.mly]: That change should be reverted, that error should be caught at scoping time (which it is).

File: src/menhir_parser.mly [hunk 3]
```diff
@@ -59,31 +60,30 @@ let translate_old_rule : old_p_rule -> p_rule = fun r ->
...
-      | P_Wrap(_)       -> fatal h.pos "Wrapping constructor in legacy rule."
-      | P_Expl(_)       -> fatal h.pos "Explicit argument in legacy rule."
+      | P_Meta(_,_)
+      | P_Patt(_,_)
+      | P_NLit(_)
+      | P_UnaO(_,_)
+      | P_BinO(_,_,_)
+      | P_Wrap(_)
+      | P_Expl(_)       -> assert false
     end;
```
Comment (rlepigre) [inline src/menhir_parser.mly]: The changes done here are not a great idea: why remove all these decent error messages? Yes they may never happen, but a nice error message is better than an assert failure. Especially in the middle of this disgusting code that is very error-prone. Plus, this has nothing to do with the current PR.

File: src/scope.ml [hunk 4]
```diff
@@ -180,9 +192,9 @@ let scope : mode -> sig_state -> env -> p_term -> tbox = fun md ss env t ->
   (* Toplevel scoping function, with handling of implicit arguments. *)
   let rec scope : env -> p_term -> tbox = fun env t ->
     (* Extract the spine. *)
-    let (p_head, args) = Syntax.get_args t in
+    let (p_head, args) = get_args t in
     (* Scope the head and obtain the implicitness of arguments. *)
-    let h = scope_head env p_head in
+    let h = scope_head env p_head args in
```
Comment (rlepigre) [inline src/scope.ml]: Here it is weird to pass an extra argument to `scope_head`. What I would do instead is add a line with something like the following (where `is_patt` is similar to `Basics.is_symb`):
```ocaml
if is_patt h && args <> [] then
  fatal t.pos "Pattern variables cannot be applied.";
```
Note that the position would be better because it would correspond to the whole application as well.

File: src/scope.ml [hunk 5]
```diff
@@ -164,6 +164,18 @@ let get_implicitness : p_term -> bool list = fun t ->
   in
   get_impl t
 
+(** [get_args t] decomposes the parser level term [t] into a spine
+   [(h,args)], when [h] is the term at the head of the application and [args]
+   is the list of all its arguments. WARNING: it does not decompose P_LLet,
+   P_UnaO, P_BinO and P_NLit. *)
```
Comment (rlepigre) [inline src/scope.ml]: The indentation of the comment is bad, why did you not keep the one we had?
Also, the constructor names are not in brackets. I suggest using the following (which avoids listing constructors because that list may evolve):
```ocaml
(** [get_args t] decomposes the parser level term [t] into a spine [(h,args)],	
    when [h] is the term at the head of the application and [args] is the list	
    of all its arguments. Note that sugared applications (e.g., infix symbols)
    are not expanded, so [h] may still be unsugared to an application. *)
```

File: src/menhir_parser.mly [hunk 6]
```diff
@@ -100,13 +100,31 @@ let translate_old_rule : old_p_rule -> p_rule = fun r ->
...
+             let qid = Pos.make p ([], prefix ^ string_of_int (i-k)) in
+             Pos.make p (P_Iden (qid, false))
+         in
+         let args = Array.init max arg in
+         let patt = Pos.make p (P_Patt(Pos.make h.pos x, args)) in
+         let rec vars i =
+           if i >= nb_exp then []
+           else Some (Pos.make p (prefix ^ string_of_int i)) :: vars (i+1)
+         in
+         Pos.make t.pos (P_Abst([vars 0, None, false], patt))
```
Comment (rlepigre) [inline src/menhir_parser.mly]: I'll review that part later, I need to open the file for that (I don't have time now).

Comment (rlepigre) [reply src/menhir_parser.mly]: I can already say that there can be some unintended variable capture here: you should use special name, not accessible to the user for the bound variables here. For example you could take `"@x"` for the prefix. Or at least check for collisions when following your naming convention (skipping the indices that are already used).

File: src/menhir_parser.mly (hunk 2 above)
Comment (fblanqui) [reply src/menhir_parser.mly]: Why this one should be caught at scoping time and not P_Prod?

File: src/menhir_parser.mly (hunk 3 above)
Comment (fblanqui) [reply src/menhir_parser.mly]: Because if this assertion fails then there is a bug. It's not a user error.

File: src/menhir_parser.mly (hunk 2 above)
Comment (rlepigre) [reply src/menhir_parser.mly]: Well, it would be best to also handle `P_Prod` at scoping time, but in that case the code for the translation is annoying to write. In any case, my main point was that these changes are not necessary.

File: src/menhir_parser.mly (hunk 3 above)
Comment (rlepigre) [reply src/menhir_parser.mly]: Well, using the legacy syntax is a user error as far as I'm concerned... :)

Frankly, it is not a big deal. I think the previous code was better, but if you want to keep it that way then at least add an `assert false` for every matched pattern. You want to known which pattern triggered the error in the case of a bug.

File: src/menhir_parser.mly [hunk 7]
```diff
@@ -40,10 +41,10 @@ let translate_old_rule : old_p_rule -> p_rule = fun r ->
     (if !verbose > 1 then Option.iter fn ao); x
   in
   let ctx = List.map get_var ctx in
-  (* Find the minimum number of arguments a variable is applied to. *)
   let is_pat_var env x =
     not (List.mem x env) && List.exists (fun y -> y.elt = x) ctx
   in
+  (* Find the maximum number of arguments a variable is applied to. *)
```
Comment (rlepigre) [inline src/menhir_parser.mly]: This comment does not make more sense here as it did above. Again, please, avoid unnecessary changes.

File: src/scope.ml (hunk 4 above)
Comment (fblanqui) [reply src/scope.ml]: Fixed in 5b0793f

File: src/scope.ml (hunk 5 above)
Comment (fblanqui) [reply src/scope.ml]: To change the indentation please push directly on my branch.

File: src/menhir_parser.mly (hunk 3 above)
Comment (fblanqui) [reply src/menhir_parser.mly]: Fixed in cd1a1fd

File: src/menhir_parser.mly (hunk 6 above)
Comment (fblanqui) [reply src/menhir_parser.mly]: You are right. I am not in favor of using "@..." because this is not parsable.

Comment (fblanqui) [reply src/menhir_parser.mly]: Should be fixed in 376bf6e

**Comment by fblanqui** [comment]
@rlepigre do you have any new comment?

**Comment by rlepigre** [comment]
> @rlepigre do you have any new comment?

I'm gonna push some simplification, but I'd like to test first. What file did you use to check that eta-expansion works?

**Comment by fblanqui** [comment]
There is tests/OK/55.dk:
```
18:59 ~/lambdapi (pattern) l tests/OK/55.dk
A : Type.
def h : (A -> A) -> (A -> A) -> A -> A.
[aa] h aa (x => aa x) --> aa.
[aa] h (x => aa x) aa --> aa.

18:59~ ~/lambdapi (pattern) lp --beautify tests/OK/55.dk
constant symbol A : TYPE
symbol h : (A ⇒ A) ⇒ (A ⇒ A) ⇒ A ⇒ A
rule h (λ v0, &aa[v0]) (λ x, &aa[x]) → λ v0, &aa[v0]
rule h (λ x, &aa[x]) (λ v0, &aa[v0]) → λ v0, &aa[v0]
```

We should probably add more.

**Comment by rlepigre** [comment]
I'm good for merging once the CI has passed.
